### PR TITLE
Fix corner case in make_match_count_matrix()

### PR DESCRIPTION
**Pull request by samuelgarcia**
* Fix corner case in make_match_count_matrix()
* Add symetric option and propagate in SymmetricSortingComparison/GroundTruthComparison
* protect with index instead of frame

### CODE REVIEW

File: src/spikeinterface/comparison/comparisontools.py [hunk 1]
```diff
@@ -232,7 +176,61 @@ def compute_matching_matrix(
...
+    1. Iterate through each spike in the first train
+    2. For each spike, find the first match in the second train.
+    3. Save the index of the first match as the new `second_train_search_start `
+    3. For each match, find as many matches as possible from the first match onwards.
+
+    An important condition here is that the same spike is not matched twice. This is managed by keeping track
+    of the last matched frame for each unit pair in `last_match_frame1` and `last_match_frame2`
+    There are corner cases where a spike can be counted twice in the the spiketrain 2 in case of bursting situations
+    (below delta_frames) in the spiketrain 1. To ensure that the number of match do not exceed the number of spike,
+    we applied a final clip.
```
Comment (zm711) [inline src/spikeinterface/comparison/comparisontools.py:228]:
```suggestion
    An important condition here is that the same spike is not matched twice. This is managed by keeping track
    of the last matched frame for each unit pair in `last_match_frame1` and `last_match_frame2`
    There are corner cases where a spike can be counted twice in the spiketrain 2 if there are bouts of bursting activity
    (below delta_frames) in the spiketrain 1. To ensure that the number of matches does not exceed the number of spikes,
    we apply a final clip.
```

File: src/spikeinterface/comparison/comparisontools.py [hunk 2]
```diff
@@ -232,7 +176,61 @@ def compute_matching_matrix(
     return compute_matching_matrix
 
 
-def make_match_count_matrix(sorting1, sorting2, delta_frames):
+def make_match_count_matrix(sorting1, sorting2, delta_frames, symmetric=False):
+    """
+    Compute a matrix representing the matches between two Sorting objects.
```
Comment (zm711) [inline src/spikeinterface/comparison/comparisontools.py:181]:
```suggestion
    Computes a matrix representing the matches between two Sorting objects.
```

File: src/spikeinterface/comparison/comparisontools.py [hunk 3]
```diff
@@ -232,7 +176,61 @@ def compute_matching_matrix(
...
+
+    Given two spike trains, this function finds matching spikes based on a temporal proximity criterion
+    defined by `delta_frames`. The resulting matrix indicates the number of matches between units
+    in `spike_frames_train1` and `spike_frames_train2` for each pair of units.
+
+    Note that this algo is not symmetric and biased toward sorting1 is the ground truth.
```
Comment (zm711) [inline src/spikeinterface/comparison/comparisontools.py:187]:
```suggestion
    Note that this algo is not symmetric and is biased with `sorting1` representing ground truth for the comparison
```

Just trying to clean this a bit. But maybe I made it too strong?

File: src/spikeinterface/comparison/comparisontools.py [hunk 4]
```diff
@@ -232,7 +176,61 @@ def compute_matching_matrix(
...
+    which signifies the minimal index in the second spike train that might match the upcoming spike
+    in the first train.
+
+    The logic can be summarized as follows:
+    1. Iterate through each spike in the first train
+    2. For each spike, find the first match in the second train.
+    3. Save the index of the first match as the new `second_train_search_start `
+    3. For each match, find as many matches as possible from the first match onwards.
+
+    An important condition here is that the same spike is not matched twice. This is managed by keeping track
```
Comment (zm711) [inline src/spikeinterface/comparison/comparisontools.py:224]:
```suggestion
    An important condition is that the same spike is not matched twice. This is managed by keeping track
```

**Review by zm711 — COMMENTED** [review]
Just a couple docstring suggestions

File: src/spikeinterface/comparison/comparisontools.py [hunk 5]
```diff
@@ -132,62 +132,6 @@ def compute_matching_matrix(
         num_units_train2,
         delta_frames,
     ):
-        """
```
Comment (h-mayorquin) [inline src/spikeinterface/comparison/comparisontools.py:135]: Why did you remove the docstring from here?

Adding the other is great but what is the gain of having less?

File: src/spikeinterface/comparison/comparisontools.py [hunk 6]
```diff
@@ -232,7 +176,61 @@ def compute_matching_matrix(
...
+    ----------
+    sorting1 : Sorting
+        An array of integer frame numbers corresponding to spike times for the first train. Must be in ascending order.
+    sorting2 : Sorting
+        An array of integer frame numbers corresponding to spike times for the second train. Must be in ascending order.
+    delta_frames : int
+        The inclusive upper limit on the frame difference for which two spikes are considered matching. That is
+        if `abs(spike_frames_train1[i] - spike_frames_train2[j]) <= delta_frames` then the spikes at `spike_frames_train1[i]`
+        and `spike_frames_train2[j]` are considered matching.
+    symmetric: bool, dfault False
```
Comment (h-mayorquin) [inline src/spikeinterface/comparison/comparisontools.py:199]: I wish we had a better name for this one.

File: src/spikeinterface/comparison/comparisontools.py (hunk 5 above)
Comment (h-mayorquin) [inline src/spikeinterface/comparison/comparisontools.py:135]: Why do we need to remove the docstring here? It is great that you added the other (my bad!) it here but is not like we are paying real state for docstrings here.

File: src/spikeinterface/comparison/comparisontools.py (hunk 6 above)
Comment (h-mayorquin) [inline src/spikeinterface/comparison/comparisontools.py:199]: I wish we could come up with a better name of this but I don't have anything better.

Comment (h-mayorquin) [reply src/spikeinterface/comparison/comparisontools.py:199]: Like, it makes me expected the returning matrix to be symmetric but that can't be.

**Review by h-mayorquin — APPROVED** [review]
Great that you solved this, @samuelgarcia .?

You can also uncomment the stuff here if you want:

https://github.com/samuelgarcia/spikeinterface/blob/dc55dfc97a66ae8681a501951bbdaa1be8342be9/src/spikeinterface/comparison/tests/test_comparisontools.py#L122-L135

It is passing locally for me.

**Review by h-mayorquin — CHANGES_REQUESTED** [review]
Yeah. Actually at least leave the arguments of the docstring if you are very adamant about real state. 

I put a lot effort writing that docstring

**Comment by zm711** [comment]
> Yeah. Actually at least leave the arguments of the docstring if you are very adamant about real state.

Agreed. It's super helpful to at least have arguments and returns in a docstring. I understand not wanting duplication or wanting to reduce lines, but knowing the ins and outs will help with development, but also with users sending in bug reports. Recently there was an issue with a function not accepting tuples when it's partner function only output tuples so knowing there was that disconnect would have found the issue earlier.

File: src/spikeinterface/comparison/comparisontools.py (hunk 5 above)
Comment (samuelgarcia) [reply src/spikeinterface/comparison/comparisontools.py:135]: Sorry, for me it was better for the higher function. This is why I moved it.
Having docstring here for function nested in another one will be never generated in the doc nor directly accessible. We can put it back maybe maybe not.

File: src/spikeinterface/comparison/comparisontools.py (hunk 6 above)
Comment (samuelgarcia) [reply src/spikeinterface/comparison/comparisontools.py:199]: `force_symmetry` ? `ensure_symmetry` ? sounds better ?

**Comment by samuelgarcia** [comment]
@h-mayorquin : sorry for the removal of the docstring. For me it was an internal function. Maybe we can put doc everywhere.

File: src/spikeinterface/comparison/comparisontools.py (hunk 6 above)
Comment (h-mayorquin) [reply src/spikeinterface/comparison/comparisontools.py:199]: Both are good choices to me actually. The two very good.

File: src/spikeinterface/comparison/comparisontools.py (hunk 5 above)
Comment (h-mayorquin) [reply src/spikeinterface/comparison/comparisontools.py:135]: That's a good point that I did not consider, but I have two points:
* It is still useful for the developer reading the code.
* I am taking your point that we should modify the way we are doing the numba functions here seriously. When we do that, I am expecting that we would be able to expose the docstrings again. And for that time, I would not like to have to urge the commit history to fetch it back. 

Given that is costing nothing, would you mind leaving it just as it is? We can discuss this after the release and I can myself go and create another PR to remove if you want it. But I don't want to get this discussion get in the way of the release and I don't want to lose the docstring to the commit history either.

Comment (samuelgarcia) [reply src/spikeinterface/comparison/comparisontools.py:135]: done

Comment (samuelgarcia) [reply src/spikeinterface/comparison/comparisontools.py:135]: OK done.

File: src/spikeinterface/comparison/comparisontools.py (hunk 1 above)
Comment (samuelgarcia) [reply src/spikeinterface/comparison/comparisontools.py]: done

File: src/spikeinterface/comparison/comparisontools.py [hunk 7]
```diff
@@ -133,11 +129,9 @@ def compute_matching_matrix(
...
     ):
         """
-        Compute a matrix representing the matches between two spike trains.
-
-        Given two spike trains, this function finds matching spikes based on a temporal proximity criterion
-        defined by `delta_frames`. The resulting matrix indicates the number of matches between units
-        in `spike_frames_train1` and `spike_frames_train2`.
+        Internal function used by `make_match_count_matrix()`.
+        This function is for one segment only.
+        The llop over segment is done in `make_match_count_matrix()`
```
Comment (alejoe91) [inline src/spikeinterface/comparison/comparisontools.py:134]:
```suggestion
        The loop over segment is done in `make_match_count_matrix()`
```

Comment (h-mayorquin) [inline src/spikeinterface/comparison/comparisontools.py]:
```suggestion
        The loop over segment is done in `make_match_count_matrix()`
```

**Review by alejoe91 — APPROVED** [review]

File: src/spikeinterface/comparison/comparisontools.py [hunk 8]
```diff
@@ -161,31 +155,8 @@ def compute_matching_matrix(
...
-        Notes
-        -----
-        This algorithm identifies matching spikes between two ordered spike trains.
-        By iterating through each spike in the first train, it compares them against spikes in the second train,
-        determining matches based on the two spikes frames being within `delta_frames` of each other.
-
-        To avoid redundant comparisons the algorithm maintains a reference, `second_train_search_start `,
-        which signifies the minimal index in the second spike train that might match the upcoming spike
-        in the first train.
+            A 2D numpy array of shape `(num_units_train1, num_units_train2)`
```
Comment (h-mayorquin) [inline src/spikeinterface/comparison/comparisontools.py:158]:
```suggestion
            A 2D numpy array of shape `(num_units_train1, num_units_train2)`. Each element `[i, j]` represents
            the count of matching spike pairs between unit `i` from `spike_frames_train1` and unit `j` from `spike_frames_train2`.
```

**Review by h-mayorquin — APPROVED** [review]
This is fine by me.

My final comment (feel free to ignore it) is that maybe the `ensure_symmetry` functionality should be in the `make_agreement_scores` as this function can't never really be symmetric and it makes more sense over there which is where you need the scores to be symmetric.
